### PR TITLE
Dense -> CSC support batch dimensions

### PR DESCRIPTION
**Pull request by amjames**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #82121
* #82122
* __->__ #83086
* #83085
* #83084

cc @alexsamardzic @nikitaved @pearu @cpuhrsch @bhosmer

### CODE REVIEW

File: aten/src/ATen/native/TensorConversions.cpp [hunk 1]
```diff
@@ -731,7 +731,78 @@ Tensor dense_to_sparse_csr(const Tensor& self) {
...
+    TORCH_CHECK(
+        not_zero_mask.sum({-2, -1}).equal(nse_per_batch),
+        "Expect the same number of specified elements per batch.");
+
+    // We need to join batches into a matrix increasing the number of cols.
+    // This allows us to create indices for col compressed matrix and de-batch
+    // them later. Avoiding a kernel start for each batch.
+    values = values.transpose(0, 1).flatten(1, 2);
+    not_zero_mask = not_zero_mask.transpose(0, 1).flatten(1, 2);
+  }
```
Comment (bhosmer) [inline aten/src/ATen/native/TensorConversions.cpp:761]: wdyt of factoring the invariant checks (#738-760) into a helper that's called from both `dense_to_sparse_csr` and `dense_to_sparse_csc`? getting common stuff out of the builder code below is more subtle but I think this acts only on the dense input and is identical (mod messaging), if I'm reading it right

File: aten/src/ATen/native/TensorConversions.cpp [hunk 2]
```diff
@@ -731,7 +731,78 @@ Tensor dense_to_sparse_csr(const Tensor& self) {
...
+        ccol_indices, n_batch, false);
+    auto batchsize_infer_last = DimVector(batch_shape);
+    // we can infer nnz/ncol+1
+    batchsize_infer_last.push_back(-1);
+    // -1 dim is nnz (per batch)
+    row_indices = row_indices.reshape(batchsize_infer_last);
+    // -1 dim is nrows (per batch) + 1
+    ccol_indices = ccol_indices.reshape(batchsize_infer_last);
+    // -1 dim is nnz (per batch)
+    values = values.reshape(batchsize_infer_last);
```
Comment (bhosmer) [inline aten/src/ATen/native/TensorConversions.cpp:796]: factoring #780-796 is more of a judgement call - personally I'd probably have an itch to try doing it as a common helper between csr and csc but not a strong opinion

**Review by bhosmer — APPROVED** [review]
Nice, couple of small factoring suggestions but up to you

File: aten/src/ATen/native/TensorConversions.cpp (hunk 1 above)
Comment (amjames) [reply aten/src/ATen/native/TensorConversions.cpp:761]: 738-754 could be factored out and shared among `dense_to_sparse_{csr,bsr,csc,bsc}` the last two lines differ subtly. `csc/bsc` require a transpose and the flatten is shifted over.

Comment (amjames) [reply aten/src/ATen/native/TensorConversions.cpp:790]: @bhosmer FYI: I refactored the handling of batch-dims on the setup (including checks) and the output side into two helpers. They have been used across the stack so the changes are on the base PR, see [this diff](https://github.com/pytorch/pytorch/pull/80781/files#diff-d4df180433a09071e891d552426911c227b30ae9b8a8e56da31046e7ecb1afbeR17-R108)

**Review by nikitaved — APPROVED** [review]
LGTM! Thank you!

**Pull request closed without merging**